### ISSUE TITLE
[FEATURE] Add source URL to a property in Frontmatter when Linking an Article

### ISSUE THREAD

**Issue by Truncated** [StrangeGirlMurph/obsidian-wikipedia-search#34]
# Description of the Problem

I use "Linking an article" to add term definitions to existing glossary pages. I've been able to add links to the source with the current template system in the body of the file, but If the url was also in a property value, I could use that to make links elsewhere with dataview.

# Description of the Solution

Within the template concept when linking an article, I would like the ability to add which property value the target URL can be added to on the page. If the property doesn't exist, then it should be added to the bottom of the frontmatter area.

# Additional

The template doesn't appear to handle frontmatter at all and the documentation for settings is a blank page, so I can't tell if there's any alternatives.

**Comment by StrangeGirlMurph** [comment]
Hey! Thanks for reaching out.

First of all: What do you mean the "documentation for settings is a blank page"? Do you mean [this](https://strangegirlmurph.github.io/obsidian-wikipedia-search/settings.html) page? It looks fine to me.

This feature seems really specific. I have to think about it. If there is a pretty and clean way to implement this option that doesn't make things unnecessarily complicated or confusing for the majority of users that don't need this feature I am probably fine with it. If you have an idea feel free to share. I don't want another field in the template options.

**Comment by Truncated** [comment]
Heyo,

**Of the First:** This is wild! Your link in the comment above works great, and I see a fully documented page with CSS functioning correctly. I have now found that the "Read the documentation!" link at the top of the settings page does the same.

I must report, however, that I was Bamboozled!

My initial experience to the documentation was through the Templates (Guide) link, shown here:
![image](https://github.com/StrangeGirlMurph/obsidian-wikipedia-search/assets/14208325/9f1e2c1d-da83-4ebc-8d1d-fa3c5a4f0c65)

Linked to this for me, with CSS of the left side bar broken overlapping the content and the Settings page consistently blank.
![image](https://github.com/StrangeGirlMurph/obsidian-wikipedia-search/assets/14208325/6b1103e8-0ad5-4850-9b27-df6f240de76c)

I can fully navigate all of the links from there and the left side bar alignment is off and the Setting page persists in being blank, no matter how internally I have tried to navigate it.

Now that I've been paying close attention, it would appear that the Guide link above is actually resolving as 404, but nearly instantly forwarding to https://strangegirlmurph.github.io/obsidian-wikipedia-search/settings.html#template-settings, but doing something extra that is breaking the CSS.

---
Given this new realization on my part and since I definitely agree that the request is rather specific,  I will first review the documentation that I now know how to uncover and see if that changes my thinking first, but I do have a few ideas to suggest.

cheers!

**Comment by StrangeGirlMurph** [comment]
Oh wow... ^^ What a weird bug! The vitepress link format changed. I fixed it with the new link and will create a new release tomorrow or the next few days with some other changes. The documentation is actually pretty good IMO 😌 Have fun!

**Comment by Truncated** [comment]
Now that I can read it, I agree that it is! Kudos on that, btw.

I'm generating a Glossary to support my research, as one does, and I have a hierarchy of definition sources that I must use to supplement these pages. Wikipedia may make up perhaps 40% of the total number of note entries and a portion of those will not have any entries when checked, so then I need to go to the next source. Additionally I need a lot more conditional options to support Iconize and Metadata Menu. As a result, I am standardizing the workflow to streamline what I can by starting with Templater.

TLDR-  Wikipedia Search cannot start this page creation party. :(

I'm omitting chained conditionals and pick lists that happen at the start which consider what note I'm currently on, what folder(s), what tag(s), yadda yadda. Eventually a page is made where a template similar to this will be triggered due to folder target: 
```
<%*
// title of the note (its value is set in the 'if' by the prompt)
let inputTitle;
if (tp.file.title.startsWith("Untitled")) {
	inputTitle = await tp.system.prompt("Enter Full Term", null, true, false);
	// Validation cycles through all separate words and capitalizes each for the Title
	glossaryTitle = inputTitle.replace(/(^\w{1})|(\s+\w{1})/g, letter => letter.toUpperCase());
	//Rename page
	await tp.file.rename(glossaryTitle);
	
	// make lowercase version of title for alias
	loweredTitle = glossaryTitle.toLowerCase();
	if (tp.user.isFirstCharLetter(loweredTitle)) {
    sentencedTitle = loweredTitle.charAt(0).toUpperCase() + loweredTitle.slice(1);
    }

	genAcronym = tp.user.craftAcronym(glossaryTitle);
	inputAcronym = await tp.system.prompt("Enter Acronym",genAcronym,false,false);
}
-%>
<% "---" %>
<% "Acronym: "+inputAcronym %>
<% "Source: " %>
<% "Definition: " %>
<% "Notional?: false" %>
<% "Category: " %>
<% "aliases: " %>
<% "   - "+inputAcronym %>
<% "   - "+loweredTitle -%>
<%* if (tp.user.isFirstCharLetter(loweredTitle)) { %>
<%* tR += "   - "+sentencedTitle -%>
<%* } %>
<% "tags: " %>
<% "Contexts: " %>
<% "Iconize: LiBookA" %>
<% "IconizeColor: " %>
<% "cssclasses: " %>
<% "---" %>

<%* app.workspace.activeLeaf.view.editor?.focus(); %>
<%* tp.hooks.on_all_templates_executed(() => {MetadataMenu.api.insertMissingFields(tp.file.path(true)); }); _%>
```
I was then using a keyboard shortcut to fire off **Link Article** with this Wikipedia Search template:
```
[{title}]({url}): {intro}
```
From Templater, I could get the auto-prompt to save me the keyboard shortcut press with this. Since I can't auto-accept results, this does save me the keypress. As far as I know, I can't pass anything from Templater to this command, and I do not believe the action can currently accept "starting text" or whatever you want to call it.
```
<%*
tp.hooks.on_all_templates_executed(() => {app.commands.executeCommandById("wikipedia-search:link-article");
});
-%>
```
### Two things that would really smooth this process out:
1. **Link Article** to open and auto search with the page title pre-filled instead of a blank field by default.
2. Enable the Wikipedia Search Settings to be able to enter values into existing property values. (which was the point of the feature title here, but with just Source).

Specifically for the example above, I would want something like:
**Definition**: {description}
**Source**: {url}

Definition is meant for short summary definition that Intro seldom is (1-2 sentences), which _sometimes_ matches up with description. As you know, it's really inconsistent so I had focused this request on the {url} which was 100%.


**For Problem 1**: adding a command that is "Link Article" but auto-populated with Page Title would be my recommendation, I think. The commands then can be chosen by upstream logic and no need to worry about a settings page element even - just use the one you want.

**For Problem 2**: IF I could trigger wikipedia-search during the Templater process in a method like tp.system.prompt(), maybe extending through [Templater's internal variable/function option](https://silentvoid13.github.io/Templater/internal-functions/contribute.html), then all the things open up and resolve themselves, I think. Is it possible that #1 with the existing Link Article could be integrated this way? If so, would that method allow for a title to be passed directly into the function for use?

Thank you for your consideration and apologies for the ramble!

**Comment by StrangeGirlMurph** [comment]
Sorry for keeping you waiting. Had a couple of busy days...
I think we/I found a solution! Regarding the first problem I am happy to add a workflow optimization setting the lets you enable/disable automatically loading the notes title as the initial search value (just like when hyperlinking) when using the link or open command. And to solve the second problem I think it's best to add another setting (not sure yet if this should be on per template basis or global) that enables replacing the template tags beyond the insert in the whole file. This might be a bit hard to understand so let me make an example:
```
---
Definition: {description}
Source: {url}
---

This is a basic markdown file.
My cursor is here: <cursor>
```

Now I use the link article command with the `[{title}]({url})` template and select the Star Wars article. With the feature enabled I would get:

```
---
Definition: American epic space opera media franchise
Source: https://en.wikipedia.org/wiki/Star_Wars
---

This is a basic markdown file.
My cursor is here: [Star Wars](https://en.wikipedia.org/wiki/Star_Wars)<cursor>
```
Normally when using an inline template you can only insert new things at the cursor position. A feature that considers all the template tags (like `{url}`) in the whole file would allow you to do exactly what you want. You create a file with templater and automatically trigger the link article command. Then the plugin automatically fills out all the tags in the current file. 

What do you think? I am a big fan. I think this is the best option that makes your workflow possible, isn't disrupting everyone elses workflow and benefits even more people.

**Comment by Truncated** [comment]
Oh, no worries!

The second option - I think that would work a treat! Let me repeat this back to you to be sure my view matches yours. I'll adapt to your example incorporated with my specific workflow.

This would be my templater file, which I have apply when I make a new file in the Star Wars folder.
```js
<% "---" %>
<% "Source: {url}" %>
<% "Definition: {description}" %>
<% "---" %>

This is a basic markdown file.
<%* app.workspace.activeLeaf.view.editor?.focus(); %>
<%* tp.hooks.on_all_templates_executed(() => {<<<wikipedia Link action here>>>}); _%>
```
The tp.hooks only fires after it's all done its work, so I won't be able to chain the Wikipedia action to "harvest" metadata variables, but I can use Templater to include the Wikipedia variables as strings, since Templater will ignore them as above.

When I create a new page using Templater, it results in a page as below:
```markdown
---
Definition: {description}
Source: {url}
---
My cursor is here: <cursor>

This is a basic markdown file.
```
The wikipedia search prompt opens immediately after this is made, and I choose the default Wikipedia Search Link Article, which gives me:

```markdown
---
Definition: American epic space opera media franchise
Source: https://en.wikipedia.org/wiki/Star_Wars
---
My cursor is here: [Star Wars](https://en.wikipedia.org/wiki/Star_Wars)
[[[block of the intro here]]] <cursor>

This is a basic markdown file.
```
Is this accurate to what you had in mind in your explanation?

**Comment by StrangeGirlMurph** [comment]
Yes exactly! I'll implement it then :)

**Comment by Truncated** [comment]
I think you are correct in your assertion of this being the best option and share your exuberance for the concept. :)

**Comment by StrangeGirlMurph** [comment]
Sorry for the long wait... I've been really busy :( I finally got to implement it now though! Check out the new update and the new docs :) Hope you like it! Thank you again for this issue.
